operator/internal/status/storage.go

```diff
@@ -25,5 +42,47 @@ func SetStorageSchemaStatus(ctx context.Context, k k8s.Client, req ctrl.Request,
 		Schemas: schemas,
 	}
 
+	for _, sc := range schemas {
+		date, err := sc.EffectiveDate.UTCTime()
+		if err != nil {
+			return kverrors.Wrap(err, "failed to parse effective date")
+		}
+		if sc.Version == lokiv1.ObjectStorageSchemaV13 && date.Before(cutoff) {
+			schemaVersionMap[sc.Version] = true
+			continue
+		} else {
+			schemaVersionMap[sc.Version] = false
+			oldSchemas = append(oldSchemas, sc)
+		}
+	}
+
+	if len(oldSchemas) > 0 {
+		if schemaVersionMap[lokiv1.ObjectStorageSchemaV13] {
+			//s.Status.Storage.Schemas[] = StorageSchemaOutOfRetention
+			if err := updateSchemaStatus(s.Status.Storage.Schemas, StorageSchemaOutOfRetention); err != nil {
+				return kverrors.Wrap(err, "error updating schema status")
+			}
+		} else {
+			//s.Status.Storage.SchemaStatus = StorageSchemaNeedsUpgrade
+			if err := updateSchemaStatus(s.Status.Storage.Schemas, StorageSchemaNeedsUpgrade); err != nil {
+				return kverrors.Wrap(err, "error updating schema status")
+			}
+
+			if err := SetWarningCondition(ctx, k, req, StorageSchemaNeedsUpgrade, lokiv1.ReasonObjectStorageSchemaShouldBeUpgraded); err != nil {
```

Comment:
I think updating the condition is better done in `status.Refresh()` where (most) of the other conditions are set. Updating the status in the schema list can still happen here, it makes sense that this happens when the rest of `.status.storage.schemas` is updated.